controller/worker/domain_migration/domain_migration.go

```diff
@@ -0,0 +1,358 @@
+		return err
+	}
+	release, err := m.client.GetAppRelease(appName)
+	if err != nil {
+		log.Error("error fetching release", "error", err)
+		return err
+	}
+	if release.Env["DEFAULT_ROUTE_DOMAIN"] != m.dm.OldDomain {
+		log.Info("already migrated")
+		return nil
+	}
+	release = dupRelease(release)
+	release.Env["DEFAULT_ROUTE_DOMAIN"] = m.dm.Domain
+	release.Env["CA_CERT"] = m.dm.TLSCert.CACert
+	if err := m.client.CreateRelease(release); err != nil {
+		log.Error("error creating release", "error", err)
+		return err
+	}
+	if err := m.client.DeployAppRelease(appName, release.ID); err != nil {
+		log.Error("error deploying release", "error", err)
+		return err
+	}
+	return nil
+}
+
+func (m *migration) maybeDeployRouter() error {
+	const appName = "router"
+	log := m.logger.New("app", appName)
+	if err := m.waitForDeployment(appName); err != nil {
+		return err
+	}
+	release, err := m.client.GetAppRelease(appName)
+	if err != nil {
+		log.Error("error fetching release", "error", err)
+		return err
+	}
+	if release.Env["TLSCERT"] != m.dm.OldTLSCert.Cert {
+		log.Info("already migrated")
+		return nil
+	}
+	release = dupRelease(release)
+	release.Env["TLSCERT"] = m.dm.TLSCert.Cert
+	release.Env["TLSKEY"] = m.dm.TLSCert.PrivateKey
+	if err := m.client.CreateRelease(release); err != nil {
+		log.Error("error creating release", "error", err)
+		return err
+	}
+	if err := m.client.DeployAppRelease(appName, release.ID); err != nil {
+		log.Error("error deploying release", "error", err)
+		return err
+	}
+	return nil
+}
+
+func (m *migration) maybeDeployDashboard() error {
+	const appName = "dashboard"
+	log := m.logger.New("app", appName)
+	if err := m.waitForDeployment(appName); err != nil {
+		return err
+	}
+	release, err := m.client.GetAppRelease(appName)
+	if err != nil {
+		log.Error("error fetching release", "error", err)
+		return err
+	}
+	if release.Env["DEFAULT_ROUTE_DOMAIN"] != m.dm.OldDomain {
+		log.Info("already migrated")
+		return nil
+	}
+	release = dupRelease(release)
+	release.Env["CA_CERT"] = m.dm.TLSCert.CACert
+	release.Env["DEFAULT_ROUTE_DOMAIN"] = m.dm.Domain
+	release.Env["CONTROLLER_DOMAIN"] = fmt.Sprintf("controller.%s", m.dm.Domain)
+	release.Env["URL"] = fmt.Sprintf("dashboard.%s", m.dm.Domain)
+	if err := m.client.CreateRelease(release); err != nil {
+		log.Error("error creating release", "error", err)
+		return err
+	}
+	if err := m.client.DeployAppRelease(appName, release.ID); err != nil {
+		log.Error("error deploying release", "error", err)
+		return err
+	}
+	return nil
+}
+
+func (m *migration) createMissingRoutes() error {
+	apps, err := m.client.AppList()
+	if err != nil {
+		return err
+	}
+	errChan := make(chan error, len(apps))
+	createMissingRoutes := func(app *ct.App) {
+		errChan <- m.appCreateMissingRoutes(app.ID)
+	}
+	for _, app := range apps {
+		go createMissingRoutes(app)
+	}
+	defer close(errChan)
+	for range apps {
+		if err := <-errChan; err != nil {
+			return err
```

Comment:
Returning an error early and thus closing the channel due to the defer could cause a panic when the remaining goroutines try to send their errors